simapp/abci.go

```diff
@@ -2,69 +2,179 @@ package simapp
-	bApp.SetVerifyVoteExtensionHandler(h.VerifyVoteExtension())
+// PrepareProposal -> Verify Prev. Block's VEs (check for 2/3+ majority ) -> Encode VEs in tx bytes
+func (app *SimApp) NewPreparePropsalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+
+		event, err := app.StakingKeeper.GetEventRecord(ctx)
+		if err != nil {
+			fmt.Println("Error occured while fetching Event Record from DB: ", err)
+		}
+		fmt.Println("Event stored in height : ", ctx.BlockHeight()-1, "event ID: ", event.Id, "event block no.: ", event.BlockNumber)
+
+		// Validate VE sigs and check whether they have 2/3+ majority
+		hasTwoThirdsSigs := true
+		if err := app.StakingKeeper.ValidateVoteExtensions(ctx, ctx.BlockHeight(), ctx.ChainID(), req.LocalLastCommit); err != nil {
+			fmt.Println("PrepareProposal: Error occured while validating VEs: ", err)
+			hasTwoThirdsSigs = false
+		}
+
+		txs := req.Txs
+		// Only include VE if there's majority
+		if hasTwoThirdsSigs {
+			fmt.Println("PrepareProposal: INJECTING VEs!")
+
+			for _, v := range req.LocalLastCommit.Votes {
+				// Appending the VEs as is. We can probably encode it (pad special bytes)
+				txs = append(txs, v.VoteExtension)
+			}
+		}
+
+		return &abci.ResponsePrepareProposal{Txs: txs}, nil
+
+	}
+
 }
 
+// Accept Proposal. Alternatively, we could check for 2/3+ VE sigs here instead of PrepareProposal
+// and reject the proposal in case we don't have a majority.
+// A downside is liveness can get negatively affected in case of invalid VEs.
+func (app *SimApp) NewProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+		// for _, tx := range req.Txs {
+		// 	// Parse the tx and ignore VEs
+		// 	decodedTx, err := app.txConfig.TxDecoder()(tx)
+		// 	if err == nil {
+		// 		if _, _, err := app.SimDeliver(app.txConfig.TxEncoder(), decodedTx); err != nil {
+		// 			return nil, fmt.Errorf("Error occured while executing tx in ProcessProposal: %v", err)
+		// 		}
+
+		// 	}
+		// 	// Set in buffer
+		// 	// if err := app.StakingKeeper.SetBuffer(ctx, *buf); err != nil {
+		// 	// 	return nil, fmt.Errorf("Error occured while setting buffer: %v", err)
+		// 	// }
+		// }
+		return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil
+	}
+
+}
+
+// ExtendVote -> Fetch data from buffer -> Vote on it accordingly -> Store result.
 func (h *VoteExtensionHandler) ExtendVote() sdk.ExtendVoteHandler {
-	return func(_ sdk.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
-		buf := make([]byte, 1024)
+	return func(ctx sdk.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+		fmt.Println("EXTENDING VOTE!", "height: ", ctx.BlockHeight())
 
-		_, err := rand.Read(buf)
+		// Fetch from buffer
+		eventRecord, err := h.app.StakingKeeper.GetEventRecordFromBuffer(ctx)
```

Comment:
If the node didn't execute processProposal, will vote SKIP.

https://docs.cometbft.com/main/spec/abci/abci++_comet_expected_behavior.html

Could change the behavior to fill the buffer here.

Reply:
Note also that the me method is called for every round. Is there any situation in which nodes would want to update their votes?

And round numbers are not available here.

Reply:
> If the node didn't execute processProposal, will vote SKIP.
> 
> https://docs.cometbft.com/main/spec/abci/abci++_comet_expected_behavior.html
> 
> Could change the behavior to fill the buffer here.
That's a good point, noted.

> Note also that the me method is called for every round. Is there any situation in which nodes would want to update their votes?
> 
> And round numbers are not available here.
Not sure which method you referred to here, but guessing that it's `ExtendVote` , we'd generally expect honest nodes to deterministically vote irrespective of the rounds but it might depend from module to module (would need to think of vulnerabilities and edge-cases per module here). We can of course have additional checks to ensure we don't get spammed with similar side-txs.